fpga_interchange/arch_pack_clusters.cc

```diff
@@ -370,6 +506,344 @@ static bool check_cluster_cells_compatibility(CellInfo *old_cell, CellInfo *new_
+    dict<CellInfo *, pool<uint32_t>, hash_ptr_ops> possible_idx;
+    for (const auto &arc : idx_to_cells)
+        for (const auto &cell : arc.second)
+            possible_idx[cell].insert(arc.first);
+    std::queue<uint32_t> prep;
+    for (const auto &arc : idx_to_cells){
+        if (arc.second.size() == 0)
+            return false;
+        if (arc.second.size()>1){
+            for (const auto &cell : arc.second){
+                auto copy_idx_to_cells(idx_to_cells);
+                copy_idx_to_cells[arc.first].clear();
+                for (uint32_t idx : possible_idx[cell]){
+                    copy_idx_to_cells[idx].erase(cell);
+                    prep.push(idx);
+                }
+                copy_idx_to_cells[arc.first].insert(cell);
+                std::queue<std::pair<uint32_t, uint32_t>> workqueue;
+                while(!prep.empty()){
+                    uint32_t idx = prep.front(); prep.pop();
+                    for (const auto &connection : cluster->connection_graph[idx].connections)
+                        if (arc.first != connection.target_idx)
+                            workqueue.push(std::pair<uint32_t, uint32_t>(arc.first, connection.target_idx));
+                }
+                binary_constraint_check(cluster, workqueue, copy_idx_to_cells, ctx);
+                if (back_solver(cluster, copy_idx_to_cells, ctx)){
+                    idx_to_cells = std::move(copy_idx_to_cells);
+                    return true;
+                }
+            }
+        }
+    }
+    return true;
+}
+
+void Arch::prepare_macro_cluster( const ClusterPOD *cluster, uint32_t index)
+{
+    Context *ctx = getCtx();
+    IdString cluster_name(cluster->name);
+
+    pool<IdString> cluster_cell_types;
+    for (auto cell_type : cluster->root_cell_types)
+        cluster_cell_types.insert(IdString(cell_type));
+
+    // Find cluster roots for each macro only ones
+    dict<IdString, CellInfo *> roots;
+    for (auto &cell : cells){
+        CellInfo *ci = cell.second.get();
+        if(ci->macro_parent == IdString())
+            continue;
+        if(ci->cluster != ClusterId())
+            continue;
+        if (!cluster_cell_types.count(ci->type))
+            continue;
+        if(roots.count(ci->macro_parent))
+            continue;
+        // Simple check based on cell type counting
+        dict<IdString, uint32_t> cells_in_macro, counter;
+        pool<IdString> cell_types;
+        for (auto &cell_type : cluster->required_cells){
+            cells_in_macro[IdString(cell_type.name)] = cell_type.count;
+            cell_types.insert(IdString(cell_type.name));
+        }
+
+        for (auto &node_cell : macro_to_cells[ci->macro_parent]){
+            auto cell_type = node_cell->type;
+            if(!counter.count(cell_type))
+                counter[cell_type] = 0;
+            counter[cell_type]++;
+            cell_types.insert(cell_type);
```

Comment:
is the separate `cell_types` pool needed, or could you just use they keys of `counter`? some comments describing what exactly these are storing would be useful

Reply:
In my opinion it's needed to store cell_types that were seen in macro_to_cells and cell_types that are used to create cluster.